flake.nix

```diff
@@ -0,0 +1,91 @@
+{
+  description = "categorifier";
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-21.11";
+    flake-utils.url = "github:numtide/flake-utils";
+    concat = {
+      url = "github:con-kitty/concat/wavewave-flake";
+      inputs.nixpkgs.follows = "nixpkgs";
+      inputs.flake-utlis.follows = "flake-utils";
+    };
+  };
+  outputs = { self, nixpkgs, flake-utils, concat }:
+    flake-utils.lib.eachSystem [ "x86_64-linux" "x86_64-darwin" ] (system:
+      let
+        parseCabalProject = import ./parse-cabal-project.nix;
+        excluded = [
+          "categorifier-vec-integration"
+          "categorifier-vec-integration-test"
+        ];
+        categorifierPackages = let parsed = parseCabalProject ./cabal.project;
+        in builtins.filter ({ name, ... }: !(builtins.elem name excluded))
+        parsed;
+        categorifierPackageNames =
+          builtins.map ({ name, ... }: name) categorifierPackages;
+        haskellOverlay = self: super:
+          builtins.listToAttrs (builtins.map ({ name, path }: {
+            inherit name;
+            value = self.callCabal2nix name (./. + "/${path}") { };
+          }) categorifierPackages);
+
+      in {
+        # This package set is only useful for CI build test.
+        # In practice, users will create a development environment composed by overlays.
+        packages = let
+          packagesOnGHC = ghcVer:
+            let
+              overlayGHC = final: prev: {
+                haskellPackages = prev.haskell.packages.${ghcVer};
+              };
+
+              newPkgs = import nixpkgs {
+                overlays = [ overlayGHC (concat.overlay.${system}) ];
+                inherit system;
+                config.allowBroken = true;
+              };
+
+              newHaskellPackages = newPkgs.haskellPackages.override (old: {
+                overrides =
+                  newPkgs.lib.composeExtensions (old.overrides or (_: _: { }))
+                  haskellOverlay;
+              });
+            in builtins.listToAttrs (builtins.map ({ name, ... }: {
+              name = ghcVer + "_" + name;
+              value = builtins.getAttr name newHaskellPackages;
+            }) categorifierPackages);
+
+        in builtins.trace (builtins.elemAt categorifierPackageNames 0)
+        (packagesOnGHC "ghc8107" // packagesOnGHC "ghc884"
+          // packagesOnGHC "ghc901" // packagesOnGHC "ghc921");
```

Comment:
Not sure where exactly to put this comment (here or in ci.yml), but we need to build against some other versions, too. E.g., we need a GHC 8.10 before 8.10.7 (and we need to keep 8.10.7), plus we need to test both GHC 9.2.1 and GHC 9.2.2. I would also like to test older versions like 8.4 and 8.6, but those are likely to be removed shortly anyway, so perhaps not worth bothering.

If we can't get all the versions we want via Nix, I'm fine with also having a smaller matrix build with just the extra versions we need to cover.

Reply:
The coverage of GHC in nixpkgs at a particular snapshot is rather limited. (just one minor version for each major GHC release, and only from GHC 8.8 to GHC 9.2 and HEAD at the moment NixOS 21.11.) This is obviously for reducing burdens of the nixpkgs maintainers. Due to the nature of nix systems, if one really wants to use unsupported compiler version, one can always use different snapshot of nixpkgs and the different systems co-exist without a trouble.

I feel like we had better support only the versions the nixpkgs maintainers supports (which already has minor fixes and custom-tailoring for individual packages) as we do not want to have burdens of maintaining all the modifications needed for dep sets for different GHC versions. That's my preference (but obvious disadvantage for nix outsiders).
Or we can make a little compilcated CI builds with different nixpkgs snapshots for the earlier versions of GHC.

But I just don't recommend supporting minor GHC versions (GHC 8.10.5 for example. too much unnecessary burden), and having non-nix build matrix, which will deter us by occasional build failures because of impure world environments, then it is likely for us to not fix the build for a long time.

Reply:
Yeah, I agree that we shouldn't test minor versions in general, but there are a couple that introduce breaking changes. In fact, I think 8.10.7 should be rescinded, as the breaking changes were reverted in 9.0.1. So we have conditionalization for 8.10.7 and for 9.2.2, so we need to test those as well as a minor version before them.

I wouldn't mind having the non-Nix matrix use `allow-failures`, but we should at least be checking. I also want a GHC-HEAD build (with allow-failures) to help us keep ahead of upcoming breakages.

Reply:
old build matrix without using nix is revived. it should always succeed regardless of buildability.